### PR TITLE
Release 4.0.0

### PR DESCRIPTION
**Pull request by erlendvollset**
Depends on
#996
#998 
#999

### CODE REVIEW

**Comment by haakonvt** [comment]
What's the timeline on this? I'm eager to get datapoints refactor into v4 ❤️  Btw, too perfect:
Screenshot 2022-08-15 at 11 57 52

**Comment by erlendvollset** [comment]
> What's the timeline on this? I'm eager to get datapoints refactor into v4 ❤️ Btw, too perfect: Screenshot 2022-08-15 at 11 57 52

Timeline is ASAP, we have other teams which have requested the envvar reworking. But there is absolutely nothing wrong with making a v5 release with your breaking datapoints changes :)

**Comment by haakonvt** [comment]
> 

Okey, sweet! 🚀

**Review by emelvarol-cognite — APPROVED** [review]
super nice

File: cognite/client/credentials.py [hunk 1]
```diff
@@ -35,6 +60,27 @@ def authorization_header(self) -> Tuple[str, str]:
 
 
 class OAuthClientCredentials(CredentialProvider):
+    """OAuth credential provider for the "Client Credentials" flow.
+
+    Args:
+        token_url (str): OAuth token url
+        client_id (str): Your application's client id.
+        client_secret (str): Your application's client secret
+        scopes (List[str]): A list of scopes.
```
Comment (emelvarol-cognite) [inline cognite/client/credentials.py:69]: I have just noticed, do we miss token_custom_args here? Or is it left out on purpose? @erlendvollset

**Pull request merged**